### ISSUE TITLE
mmv doesn't cope with directory renames correctly

### ISSUE THREAD

**Issue by rogpeppe** [itchyny/mmv#11]
Firstly, thanks for a great idea! I really like the idea of this command and I'll probably end up using it quite a bit.
One use case I thought for it was for rearranging entire directory hierarchies, e.g. `mmv $(find .)`

However, it doesn't seem to understand that moving a directory also moves the contents of that directory, including
possible sources of the rename operation.

Here's a demonstration of this issue, scripted using the [testscript command](https://pkg.go.dev/github.com/rogpeppe/go-internal/cmd/testscript?readme=expanded#section-readme):

```
env EDITOR=$WORK/mmv-script-edit
chmod 755 $EDITOR
exec cp -r d orig
exec mmv d/a d/b d/b/a d/a/c d/a/b
cmp d/b/untouched orig/a/untouched
cmp d/a/untouched orig/b/untouched
cmp d/a/a orig/b/a
cmp d/d/c orig/d/a/c
cmp d/c/c orig/d/a/b
-- mmv-script-edit --
#!/bin/sh
cp script $1
-- script --
d/b
d/a
d/a/a
d/d/c
d/c/c
-- d/a/untouched --
a-untouched
-- d/b/untouched
b-untouched
-- d/b/a --
b-a
-- d/a/c --
a-c
-- d/a/b --
a-b
```

Running this script generates a sequence of renames like this:

```
[{d/b d/5577006791947779410} {d/a d/b} {d/5577006791947779410 d/a} {d/b/a d/a/a} {d/a/c d/d/c} {d/a/b d/c/c}]
```
The fourth rename of `d/b/a` to `d/a/a` fails because `d/b` has previously been renamed to `d/a`.

The `mmv` algorithm needs to be somewhat more sophisticated. On the positive side, fixing this will make `mmv` into an even more indispensable command!

**Comment by itchyny** [comment]
I didn't expect this case to happen and I want to change to report an error.
~The mmv command does not expect tree structure in the argument files.~
It successfully moves files in a directory into directories, or files in directories to a directory but no more complex renames.

The following case renames the directory, but the user may keep the a/x as it is or not?
```
a   -> b
a/x -> a/x
```

The file y should be in a/ or b/ in the following case? I prefer error rather than implementing one of them.
```
a   -> b
a/x -> a/y
```

The following case fails.
```
a   -> a/x
a/x -> a
```

**Comment by rogpeppe** [comment]
I think the invariant should be that all source files end up in their chosen destinations, regardless of structure. Moving a directory should implicitly move all files inside it unless they're explicitly named.

I think that maps well to the intent of this tool and the semantics that have already been chosen with respect to swapping file names, for example.

So in your first case, a/x should end up in a/x, but the rest of the contents of a should end up in b.

Similarly in your second case, a/x should end up in a/y while the rest of the files in a should end up in b.

The last case should only fail if a/x isn't a directory, otherwise it seems ok to me.

Implementing this is an interesting algorithmic challenge, but should be worth it IMO.

**Comment by itchyny** [comment]
I'll implement this in this weekend.

**Comment by rogpeppe** [comment]
I've been thinking about this a bit. A basic sketch for a sufficient (but somewhat inefficient) algorithm might be something like this:

- stat each source file, mark it with the file's type (file or directory) and check (using [os.SameFile](https://golang.org/pkg/os/#SameFile)?) that they're all unique. 
- check that the mapping from source to destination is one-to-one and onto
- create a set of all the mappings from source to destination, using absolute path names with symlinks evaluated..
- for each destination, if there are any other destinations that are ancestors, ensure that they were marked as directories
- for each source, move it to a random temporary name in one of its parent directories - the deepest parent directory that isn't a within any other name mentioned in any source or destination.
- create a list of all the mappings ordered by number of path elements, fewest first
- for each mapping in the above list, move its temporary file to the destination, creating directories if needed

The above algorithm will move every file twice, but this can be avoided by avoiding the temporary move for any file for which no source or destination path is a child or ancestor.

The big wrinkle here is symbolic links. I suspect they're going to a huge pain, but I haven't managed to think that through yet. Maybe just leave proper handling of symbolic links as a TODO for now... :)

**Comment by itchyny** [comment]
I finally implemented this feature, thank you for feedback.

**Comment by itchyny** [comment]
I finally implemented this feature, thank you for feedback.

**Comment by itchyny** [comment]
Released by [v0.1.3](https://github.com/itchyny/mmv/releases/tag/v0.1.3).
